pallets/auctions/src/lib.rs

```diff
@@ -160,171 +154,297 @@ pub mod pallet {
 			let sender = ensure_signed(origin)?;
-			let new_auction_id = Self::new_auction(auction_info)?;
-			Self::deposit_event(Event::AuctionCreated(sender, new_auction_id));
-			Ok(().into())
+
+			match &auction {
+				Auction::English(auction_object) => {
+					Self::validate_general_data(&auction_object.general_data)?;
+					Self::validate_create(&auction_object.general_data)?;
+					Self::handle_create(sender, &auction, &auction_object.general_data)?;
+				}
+			}
+
+			Ok(())
 		}
 
-		#[pallet::weight(<T as Config>::WeightInfo::bid_value())]
-		pub fn bid_value(origin: OriginFor<T>, id: T::AuctionId, value: BalanceOf<T>) -> DispatchResultWithPostInfo {
+		#[pallet::weight(<T as Config>::WeightInfo::update_auction())]
+		pub fn update(origin: OriginFor<T>, id: T::AuctionId, updated_auction: Auction<T>) -> DispatchResult {
 			let sender = ensure_signed(origin)?;
 
-			Self::bid(sender.clone(), id, value)?;
-			Self::deposit_event(Event::Bid(id, sender, value));
-			Ok(().into())
+			match &updated_auction {
+				Auction::English(auction_object) => {
+					Self::validate_general_data(&auction_object.general_data)?;
+					Self::handle_update(sender, id, updated_auction.clone(), &auction_object.general_data)?;
+				}
+			}
+
+			Ok(())
 		}
 
-		#[pallet::weight(<T as Config>::WeightInfo::delete_auction())]
-		pub fn delete_auction(origin: OriginFor<T>, id: T::AuctionId) -> DispatchResultWithPostInfo {
-			let _sender = ensure_signed(origin)?;
+		#[pallet::weight(<T as Config>::WeightInfo::destroy_auction())]
+		pub fn destroy(origin: OriginFor<T>, id: T::AuctionId) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let auction = <Auctions<T>>::get(id).ok_or(Error::<T>::AuctionNotExist)?;
+
+			match &auction {
+				Auction::English(auction_object) => {
+					Self::validate_update(sender, &auction_object.general_data)?;
+					Self::handle_destroy(id, &auction_object.general_data)?;
+				}
+			}
 
-			Self::remove_auction(id)?;
-			Self::deposit_event(Event::AuctionRemoved(id));
-			Ok(().into())
+			Ok(())
 		}
-	}
 
-	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
-		fn on_finalize(now: T::BlockNumber) {
-			Self::conclude_auction(now);
+		#[pallet::weight(<T as Config>::WeightInfo::bid())]
+		pub fn bid(origin: OriginFor<T>, auction_id: T::AuctionId, value: BalanceOf<T>) -> DispatchResult {
+			let bidder = ensure_signed(origin)?;
+
+			let bid = <Auctions<T>>::try_mutate(auction_id, |maybe_auction| -> DispatchResult {
+				let auction = maybe_auction.as_mut().ok_or(Error::<T>::AuctionNotExist)?;
+
+				match auction {
+					Auction::English(auction_object) => {
+						Self::validate_bid(&bidder, &auction_object.general_data, value)?;
+						auction_object.bid(bidder.clone(), value)?;
+					}
+				}
+
+				Ok(())
+			});
+
+			if let Err(err) = bid {
+				return Err(err);
+			}
+
+			Self::deposit_event(Event::Bid(auction_id, bidder, value));
+
+			Ok(())
```

Comment:
You can simplify this by moving the deposit event insist the try mutate and return the value of that.

```suggestion
			<Auctions<T>>::try_mutate(auction_id, |maybe_auction| -> DispatchResult {
				let auction = maybe_auction.as_mut().ok_or(Error::<T>::AuctionNotExist)?;

				match auction {
					Auction::English(auction_object) => {
						Self::validate_bid(&bidder, &auction_object.general_data, value)?;
						auction_object.bid(bidder.clone(), value)?;
					}
				}
                                Self::deposit_event(Event::Bid(auction_id, bidder, value));
				Ok(())
			})
```